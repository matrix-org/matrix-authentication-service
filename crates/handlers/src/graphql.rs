// Copyright 2022 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use async_graphql::{
    extensions::{ApolloTracing, Tracing},
    http::{playground_source, GraphQLPlaygroundConfig, MultipartOptions},
};
use axum::{
    async_trait,
    extract::{BodyStream, RawQuery, State},
    response::{Html, IntoResponse},
    Json, TypedHeader,
};
use axum_extra::extract::PrivateCookieJar;
use futures_util::TryStreamExt;
use headers::{ContentType, HeaderValue};
use hyper::header::CACHE_CONTROL;
use mas_axum_utils::{FancyError, SessionInfoExt};
use mas_graphql::{Requester, Schema};
use mas_keystore::Encrypter;
use mas_storage::{BoxClock, BoxRepository, BoxRng, Repository, RepositoryError, SystemClock};
use mas_storage_pg::PgRepository;
use rand::{thread_rng, SeedableRng};
use rand_chacha::ChaChaRng;
use sqlx::PgPool;
use tracing::{info_span, Instrument};

struct GraphQLState {
    pool: PgPool,
}

#[async_trait]
impl mas_graphql::State for GraphQLState {
    async fn repository(&self) -> Result<BoxRepository, RepositoryError> {
        let repo = PgRepository::from_pool(&self.pool)
            .await
            .map_err(RepositoryError::from_error)?;

        Ok(repo.map_err(RepositoryError::from_error).boxed())
    }

    fn clock(&self) -> BoxClock {
        let clock = SystemClock::default();
        Box::new(clock)
    }

    fn rng(&self) -> BoxRng {
        #[allow(clippy::disallowed_methods)]
        let rng = thread_rng();

        let rng = ChaChaRng::from_rng(rng).expect("Failed to seed rng");
        Box::new(rng)
    }
}

#[must_use]
pub fn schema(pool: &PgPool) -> Schema {
    let state = GraphQLState { pool: pool.clone() };
    let state: mas_graphql::BoxState = Box::new(state);

    mas_graphql::schema_builder()
        .extension(Tracing)
        .extension(ApolloTracing)
        .data(state)
        .finish()
}

fn span_for_graphql_request(request: &async_graphql::Request) -> tracing::Span {
    let span = info_span!(
        "GraphQL operation",
        "otel.name" = tracing::field::Empty,
        "otel.kind" = "server",
        "graphql.document" = request.query,
        "graphql.operation.name" = tracing::field::Empty,
    );

    if let Some(name) = &request.operation_name {
        span.record("otel.name", name);
        span.record("graphql.operation.name", name);
    }

    span
}

pub async fn post(
    State(schema): State<Schema>,
    mut repo: BoxRepository,
    cookie_jar: PrivateCookieJar<Encrypter>,
    content_type: Option<TypedHeader<ContentType>>,
    body: BodyStream,
) -> Result<impl IntoResponse, FancyError> {
    let (session_info, _cookie_jar) = cookie_jar.session_info();
    let maybe_session = session_info.load_session(&mut repo).await?;
    let requester = Requester::from(maybe_session);
    repo.cancel().await?;

    let content_type = content_type.map(|TypedHeader(h)| h.to_string());

    let request = async_graphql::http::receive_body(
        content_type,
        body.map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))
            .into_async_read(),
        MultipartOptions::default(),
    )
    .await?
    .data(requester); // XXX: this should probably return another error response?

    let span = span_for_graphql_request(&request);
    let response = schema.execute(request).instrument(span).await;

    let cache_control = response
        .cache_control
        .value()
        .and_then(|v| HeaderValue::from_str(&v).ok())
        .map(|h| [(CACHE_CONTROL, h)]);

    let headers = response.http_headers.clone();

    Ok((headers, cache_control, Json(response)))
}

pub async fn get(
    State(schema): State<Schema>,
    mut repo: BoxRepository,
    cookie_jar: PrivateCookieJar<Encrypter>,
    RawQuery(query): RawQuery,
) -> Result<impl IntoResponse, FancyError> {
    let (session_info, _cookie_jar) = cookie_jar.session_info();
    let maybe_session = session_info.load_session(&mut repo).await?;
    let requester = Requester::from(maybe_session);
    repo.cancel().await?;

    let request =
        async_graphql::http::parse_query_string(&query.unwrap_or_default())?.data(requester);

    let span = span_for_graphql_request(&request);
    let response = schema.execute(request).instrument(span).await;

    let cache_control = response
        .cache_control
        .value()
        .and_then(|v| HeaderValue::from_str(&v).ok())
        .map(|h| [(CACHE_CONTROL, h)]);

    let headers = response.http_headers.clone();

    Ok((headers, cache_control, Json(response)))
}

pub async fn playground() -> impl IntoResponse {
    Html(playground_source(
        GraphQLPlaygroundConfig::new("/graphql").with_setting("request.credentials", "include"),
    ))
}
