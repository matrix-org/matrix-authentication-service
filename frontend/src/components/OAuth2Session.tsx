// Copyright 2022 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { Button } from "@vector-im/compound-web";
import { atom, useSetAtom } from "jotai";
import { atomFamily } from "jotai/utils";
import { atomWithMutation } from "jotai-urql";
import { useTransition } from "react";

import { FragmentType, graphql, useFragment } from "../gql";
import { getDeviceIdFromScope } from "../utils/deviceIdFromScope";
// import LoadingSpinner from "./LoadingSpinner/LoadingSpinner";
import { Session } from "./Session/Session";

const FRAGMENT = graphql(/* GraphQL */ `
  fragment OAuth2Session_session on Oauth2Session {
    id
    scope
    createdAt
    finishedAt
    client {
      id
      clientId
      clientName
      clientUri
    }
  }
`);

type Oauth2SessionType = {
  id: string;
  scope: string;
  createdAt: number;
  finishedAt?: number;
  client: {
    id: string;
    clientId: string;
    clientName: string;
    clientUri: string;
  }
}

const END_SESSION_MUTATION = graphql(/* GraphQL */ `
  mutation EndOAuth2Session($id: ID!) {
    endOauth2Session(input: { oauth2SessionId: $id }) {
      status
      oauth2Session {
        id
        ...OAuth2Session_session
      }
    }
  }
`);

const endSessionFamily = atomFamily((id: string) => {
  const endSession = atomWithMutation(END_SESSION_MUTATION);

  // A proxy atom which pre-sets the id variable in the mutation
  const endSessionAtom = atom(
    (get) => get(endSession),
    (get, set) => set(endSession, { id }),
  );

  return endSessionAtom;
});

type Props = {
  session: FragmentType<typeof FRAGMENT>;
};

const OAuth2Session: React.FC<Props> = ({ session }) => {
  const [pending, startTransition] = useTransition();
  const data = useFragment<Oauth2SessionType>(FRAGMENT, session);
  const endSession = useSetAtom(endSessionFamily(data.id));

  // @TODO(kerrya) make this wait for session refresh properly
  // https://github.com/matrix-org/matrix-authentication-service/issues/1533
  const onSessionEnd = (): void => {
    startTransition(() => {
      endSession();
    });
  };

  const sessionName = getDeviceIdFromScope(data.scope);

  return (
    <Session
      id={data.id}
      name={sessionName}
      createdAt={data.createdAt}
      finishedAt={data.finishedAt}
      clientName={data.client.clientName}
    >
      {!data.finishedAt && (
        <Button
          kind="destructive"
          size="sm"
          onClick={onSessionEnd}
          disabled={pending}
        >
          {/* @TODO(kerrya) put this back after pending state works properly */}
          {/* { pending && <LoadingSpinner />} */}
          End session
        </Button>
      )}
    </Session>
  );
};

export default OAuth2Session;
