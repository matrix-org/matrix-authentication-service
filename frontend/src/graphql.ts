// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { createClient, fetchExchange } from "@urql/core";
import { devtoolsExchange } from "@urql/devtools";
import { cacheExchange } from "@urql/exchange-graphcache";
import { refocusExchange } from "@urql/exchange-refocus";
import { requestPolicyExchange } from "@urql/exchange-request-policy";

import appConfig from "./config";
import type {
  MutationAddEmailArgs,
  MutationRemoveEmailArgs,
  MutationVerifyEmailArgs,
  RemoveEmailPayload,
  VerifyEmailPayload,
} from "./gql/graphql";
import schema from "./gql/schema";

const cache = cacheExchange({
  schema,
  keys: {
    // This is embedded in the `User` entity and has no ID
    MatrixUser: () => null,
  },
  updates: {
    Mutation: {
      addEmail: (result, args: MutationAddEmailArgs, cache, _info) => {
        const key = cache.keyOfEntity({
          __typename: "User",
          id: args.input.userId,
        });

        // Invalidate the emails field on the User object so that it gets refetched
        cache
          .inspectFields(key)
          .filter((field) => field.fieldName === "emails")
          .forEach((field) => {
            cache.invalidate(key, field.fieldName, field.arguments);
          });
      },

      removeEmail: (
        result: { removeEmail?: RemoveEmailPayload },
        args: MutationRemoveEmailArgs,
        cache,
        _info,
      ) => {
        // Invalidate the email entity
        cache.invalidate({
          __typename: "UserEmail",
          id: args.input.userEmailId,
        });

        // Let's try to figure out the userId to invalidate the emails field on the User object
        const userId = result.removeEmail?.user?.id;
        if (userId) {
          const key = cache.keyOfEntity({
            __typename: "User",
            id: userId,
          });

          // Invalidate the emails field on the User object so that it gets refetched
          cache
            .inspectFields(key)
            .filter((field) => field.fieldName === "emails")
            .forEach((field) => {
              cache.invalidate(key, field.fieldName, field.arguments);
            });
        }
      },

      verifyEmail: (
        result: { verifyEmail?: VerifyEmailPayload },
        args: MutationVerifyEmailArgs,
        cache,
        _info,
      ) => {
        // Invalidate the email entity
        cache.invalidate({
          __typename: "UserEmail",
          id: args.input.userEmailId,
        });

        // Let's try to figure out the userId to invalidate the emails field on the User object
        const userId = result.verifyEmail?.user?.id;
        if (userId) {
          const key = cache.keyOfEntity({
            __typename: "User",
            id: userId,
          });

          // Invalidate the emails field on the User object so that it gets refetched
          cache
            .inspectFields(key)
            .filter((field) => field.fieldName === "emails")
            .forEach((field) => {
              cache.invalidate(key, field.fieldName, field.arguments);
            });
        }
      },
    },
  },
});

const exchanges = [
  // This sets the policy to 'cache-and-network' after 5 minutes
  requestPolicyExchange({
    ttl: 1000 * 60 * 5, // 5 minute
  }),

  // This refetches all queries when the tab is refocused
  refocusExchange(),

  // The unified cache
  cache,

  // Use `fetch` to execute the requests
  fetchExchange,
];

export const client = createClient({
  url: appConfig.graphqlEndpoint,
  suspense: true,
  // Add the devtools exchange in development
  exchanges: import.meta.env.DEV ? [devtoolsExchange, ...exchanges] : exchanges,
});
