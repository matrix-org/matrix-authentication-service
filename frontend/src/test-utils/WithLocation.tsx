// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// @vitest-environment happy-dom

import { Provider } from "jotai";
import { useHydrateAtoms } from "jotai/utils";

import { appConfigAtom, locationAtom } from "../routing";

const HydrateLocation: React.FC<React.PropsWithChildren<{ path: string }>> = ({
  children,
  path,
}) => {
  useHydrateAtoms([
    [appConfigAtom, { root: "/" }],
    [locationAtom, { pathname: path }],
  ]);
  return <>{children}</>;
};

/**
 * Utility for testing components that rely on routing or location
 * For example any component that includes a <Link />
 * Eg:
 * ```
 * const component = create(
      <WithLocation path="/">
        <NavItem route={{ type: "profile" }}>Active</NavItem>
      </WithLocation>,
    );
 * ```
 */
export const WithLocation: React.FC<
  React.PropsWithChildren<{ path?: string }>
> = ({ children, path }) => {
  return (
    <Provider>
      <HydrateLocation path={path || "/"}>{children}</HydrateLocation>
    </Provider>
  );
};
