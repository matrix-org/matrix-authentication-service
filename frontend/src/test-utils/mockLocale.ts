// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { vi } from "vitest";

/**
 * Mock the locale on Intl.DateTimeFormat
 * To achieve stable formatted dates across environments
 * Defaults to `en-GB`
 */
export const mockLocale = (defaultLocale = "en-GB"): void => {
  const { DateTimeFormat } = Intl;
  vi.spyOn(Intl, "DateTimeFormat").mockImplementation(
    (
      locales?: Intl.LocalesArgument,
      options?: Intl.DateTimeFormatOptions | undefined,
    ) => new DateTimeFormat(locales || defaultLocale, options),
  );
};
