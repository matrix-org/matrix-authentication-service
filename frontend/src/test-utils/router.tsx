// Copyright 2024 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {
  RouterProvider,
  createMemoryHistory,
  createRootRoute,
  createRoute,
  createRouter,
} from "@tanstack/react-router";

const rootRoute = createRootRoute();
const index = createRoute({ getParentRoute: () => rootRoute, path: "/" });
rootRoute.addChildren([index]);

const router = createRouter({
  history: createMemoryHistory(),
  routeTree: rootRoute,
});

const routerReady = router.load();
// Because we also use this in the stories, we need to make sure we call this only in tests
if (import.meta.vitest) {
  // Make sure the router is ready before running any tests
  import("vitest").then(({ beforeAll }) =>
    beforeAll(async () => await routerReady),
  );
}

export const DummyRouter: React.FC<React.PropsWithChildren> = ({
  children,
}) => (
  /** @ts-expect-error: The router we pass doesn't match the "real" router, which is fine for tests */
  <RouterProvider router={router} defaultComponent={() => children} />
);
