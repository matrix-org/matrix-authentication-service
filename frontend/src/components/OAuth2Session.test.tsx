// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// @vitest-environment happy-dom

import { create } from "react-test-renderer";
import { Provider } from "urql";
import { describe, expect, it, beforeAll } from "vitest";
import { never } from "wonka";

import { makeFragmentData } from "../gql";
import { Oauth2ApplicationType } from "../gql/graphql";
import { mockLocale } from "../test-utils/mockLocale";
import { DummyRouter } from "../test-utils/router";

import OAuth2Session, { FRAGMENT } from "./OAuth2Session";

describe("<OAuth2Session />", () => {
  const mockClient = {
    executeQuery: (): typeof never => never,
  };

  const defaultSession = {
    id: "session-id",
    scope:
      "openid urn:matrix:org.matrix.msc2967.client:api:* urn:matrix:org.matrix.msc2967.client:device:abcd1234",
    createdAt: "2023-06-29T03:35:17.451292+00:00",
    lastActiveIp: "1.2.3.4",
    client: {
      id: "test-id",
      clientId: "test-client-id",
      clientName: "Element",
      clientUri: "https://element.io",
      applicationType: Oauth2ApplicationType.Web,
    },
  };

  const finishedAt = "2023-06-29T03:35:19.451292+00:00";

  beforeAll(() => mockLocale());

  it("renders an active session", () => {
    const session = makeFragmentData(defaultSession, FRAGMENT);

    const component = create(
      <Provider value={mockClient}>
        <DummyRouter>
          <OAuth2Session session={session} />
        </DummyRouter>
      </Provider>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("renders a finished session", () => {
    const session = makeFragmentData(
      {
        ...defaultSession,
        finishedAt,
      },
      FRAGMENT,
    );
    const component = create(
      <Provider value={mockClient}>
        <DummyRouter>
          <OAuth2Session session={session} />
        </DummyRouter>
      </Provider>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("renders correct icon for a native session", () => {
    const session = makeFragmentData(
      {
        ...defaultSession,
        finishedAt,
        client: {
          ...defaultSession.client,
          applicationType: Oauth2ApplicationType.Native,
        },
      },
      FRAGMENT,
    );
    const component = create(
      <Provider value={mockClient}>
        <DummyRouter>
          <OAuth2Session session={session} />
        </DummyRouter>
      </Provider>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
