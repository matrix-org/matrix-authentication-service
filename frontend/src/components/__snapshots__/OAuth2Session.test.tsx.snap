// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<OAuth2Session /> > renders a finished session 1`] = `
<div
  className="_block_17898c"
>
  <h6
    className="_font-body-md-semibold_1jx6b_64 _sessionName_634806"
    title="session-id"
  >
    <a
      className=""
      href="/session/abcd1234"
      onClick={[Function]}
    >
      abcd1234
    </a>
  </h6>
  <p
    className="_font-body-sm-semibold_1jx6b_45 _sessionMetadata_634806"
  >
    Signed in 
    <code>
      2023-06-29T03:35:17.451292+00:00
    </code>
  </p>
  <p
    className="_font-body-sm-semibold_1jx6b_45 _sessionMetadata_634806"
    data-finished={true}
  >
    Finished 
    <code>
      2023-06-29T03:35:19.451292+00:00
    </code>
  </p>
  <p
    className="_font-body-sm-regular_1jx6b_40 _sessionMetadata_634806"
  >
    Client:
     
    <span
      className="_font-body-sm-semibold_1jx6b_45 _sessionMetadata_634806"
    >
      Element
    </span>
  </p>
</div>
`;

exports[`<OAuth2Session /> > renders an active session 1`] = `
<div
  className="_block_17898c"
>
  <h6
    className="_font-body-md-semibold_1jx6b_64 _sessionName_634806"
    title="session-id"
  >
    <a
      className=""
      href="/session/abcd1234"
      onClick={[Function]}
    >
      abcd1234
    </a>
  </h6>
  <p
    className="_font-body-sm-semibold_1jx6b_45 _sessionMetadata_634806"
  >
    Signed in 
    <code>
      2023-06-29T03:35:17.451292+00:00
    </code>
  </p>
  <p
    className="_font-body-sm-regular_1jx6b_40 _sessionMetadata_634806"
  >
    Client:
     
    <span
      className="_font-body-sm-semibold_1jx6b_45 _sessionMetadata_634806"
    >
      Element
    </span>
  </p>
  <div
    className="_sessionActions_634806"
  >
    <button
      className="_button_1raud_17"
      data-kind="destructive"
      data-size="sm"
      disabled={false}
      onClick={[Function]}
      role="button"
      tabIndex={0}
    >
      End session
    </button>
  </div>
</div>
`;
